node/src/archive/transformer.rs

```diff
@@ -0,0 +1,158 @@
+) {
+    // 1st Group events by origin (TxHash) & throw away the ones that
+    // don't have an origin
+    let mut moonlight_is_already_grouped: BTreeMap<TxHash, Vec<ContractEvent>> =
+        BTreeMap::new();
+    for event in block_events {
+        if let Some(origin) = event.origin {
+            let event_to_analyze = event.event;
+
+            moonlight_is_already_grouped
+                .entry(origin)
+                .or_default()
+                .push(event_to_analyze);
+        } else {
+            continue;
+        };
+    }
+
+    // 2nd Keep only the event groups which contain a moonlight in-
+    // or outflow
+
+    let mut address_mappings: Vec<(AccountPublicKey, TxHash)> = vec![];
+    let mut memo_mappings: Vec<(Vec<u8>, TxHash)> = vec![];
+    let mut moonlight_tx_groups = vec![];
+    // Iterate over the grouped events and push them to the groups vector in
+    // the new format if they are moonlight events
+    for (tx_hash, group) in moonlight_is_already_grouped {
+        let is_moonlight = group.iter().any(|event| {
+            // Make sure that the events originate from the transfer contract.
+            if event.target.0 != TRANSFER_CONTRACT {
+                return false;
+            }
+
+            /*
+            Cases of a moonlight in- or outflow:
+            1. Any MoonlightTransactionEvent. This implicitly also catches a moonlight outflow for deposit or convert (from moonlight)
+            2a. Any withdraw event where the receiver is moonlight. (from phoenix)
+            2b. Any mint event where the receiver is moonlight. (from staking)
+            3. Any convert event where the receiver is moonlight. (from phoenix)
+            */
+            match event.topic.as_str() {
+                MOONLIGHT_TOPIC => {
+                    /*
+                        This also catches deposits & converts.
+                        For deposits & convert the sender will be Some(pk) where pk is the same as the from field of the MoonlightTransactionEvent
+                    */
+                    if let Ok(moonlight_event) =
+                        rkyv::from_bytes::<MoonlightTransactionEvent>(
+                            &event.data,
+                        )
+                    {
+                        address_mappings
+                            .push((moonlight_event.sender, tx_hash));
+                        if let Some(receiver) = moonlight_event.receiver {
+                            if moonlight_event.sender != receiver {
+                                // don't push if tx is sent to self
+                                address_mappings.push((receiver, tx_hash));
+                            }
+                        }
+
+                        if !moonlight_event.memo.is_empty() {
+                            memo_mappings.push((moonlight_event.memo, tx_hash));
+                        }
+
+                        return true;
+                    }
+                    false
```

Comment:
I find the mix of 'return' and direct values confusing. Could this be unified? For example returns eliminated?